File: MyFbApp/MyFbAppFSharp/DatabaseManager.fs
namespace MyFbAppFSharp

open SQLite
open System
open System.IO
open DatabaseModels
open System.Linq.Expressions

module databaseManager =

    type FacebookPostsDbObject() =
        [<PrimaryKey; AutoIncrement>]
        member val Id = 0 with get, set
        member val TimeStamp = DateTime.MinValue with get, set
        member val Message = "" with get, set
        member val Created_time = DateTime.MinValue with get, set
        member val PostsId = "" with get, set
        member val CommentsNumber = 0 with get, set
        member val UserId = "" with get, set

    let convertFbPostDbToObject (item: FacebookPostsDb) =
        let obj = FacebookPostsDbObject()
        obj.Id <- item.Id
        obj.TimeStamp <- item.TimeStamp
        obj.Message <- item.Message
        obj.Created_time <- item.Created_time
        obj.PostsId <- item.PostsId
        obj.CommentsNumber <- item.CommentsNumber
        obj.UserId <- item.UserId
        obj

    let convertFbPostDbToModel (obj: FacebookPostsDbObject) : FacebookPostsDb =
        { Id = obj.Id
          TimeStamp = obj.TimeStamp
          Message = obj.Message
          Created_time = obj.Created_time
          PostsId = obj.PostsId
          CommentsNumber = obj.CommentsNumber
          UserId = obj.UserId }

    type FacebookProfileDbObject() =
        [<PrimaryKey; AutoIncrement>]
        member val Id = 0 with get, set
        member val TimeStamp = DateTime.MinValue with get, set
        member val Name = "" with get, set
        member val UserId = "" with get, set

    let convertFbProfileDbToObject (item: FacebookProfileDb) =
        let obj = FacebookProfileDbObject()
        obj.Id <- item.Id
        obj.TimeStamp <- item.TimeStamp
        obj.Name <- item.Name
        obj.UserId <- item.UserId
        obj

    let convertFbProfileDbToModel (obj: FacebookProfileDbObject) : FacebookProfileDb =
        { Id = obj.Id
          TimeStamp = obj.TimeStamp
          Name = obj.Name
          UserId = obj.UserId }

    type TokenDbObject() =
        [<PrimaryKey; AutoIncrement>]
        member val Id = 0 with get, set
        member val LongLiveToken = "" with get, set

    let convertTokenDbToObject (item: TokenDb) =
        let obj = TokenDbObject()
        obj.Id <- item.Id
        obj.LongLiveToken <- item.LongLiveToken
        obj

    let convertTokenDbToModel (obj: TokenDbObject) : TokenDb =
        { Id = obj.Id
          LongLiveToken = obj.LongLiveToken }

    let dbPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.Personal), "MyDB.db3")

    let connect =
        let db = new SQLiteConnection(dbPath)
        db

    let database = connect

    let CreateTables =
        do database.CreateTable<FacebookPostsDbObject>() |> ignore
           database.CreateTable<FacebookProfileDbObject>() |> ignore
           database.CreateTable<TokenDbObject>() |> ignore

    let CheckToken = 
        let count = database.Table<TokenDbObject>().Count()
        count > 0

    let getLongLiveToken = 
        database.Table<TokenDbObject>().First().LongLiveToken

    let UpdateLongLiveToken token = 
        do database.DeleteAll<TokenDbObject>() |> ignore
        let newToken = TokenDbObject(Id= 0, LongLiveToken= token)
        (*let obj = convertTokenDbToObject token*)
        do database.Insert(newToken) |> ignore


    let UpdateProfile profile =
        let obj = convertFbProfileDbToObject profile
        do database.Execute("DELETE FROM FacebookProfileDb WHERE UserId = ?", obj.Id) |> ignore
           database.Insert(obj) |> ignore

    let UpdatePosts posts = 
        do for post in posts do
              let obj = convertFbPostDbToObject post
              do database.Update(obj) |> ignore

    let getFacebookProfileDb = 
        database.Table<FacebookProfileDbObject>().First()

    let getFacebookPostsDb = 
        database.Table<FacebookPostsDbObject>().ToList()

    (*let CheckTimeStamp (userId: string) =
        let result = DateTime.Compare(database.Table<FacebookProfileDbObject>().Where(w => w.UserId == userId).First().TimeStamp, DateTime.Now.AddDays(-5.0)) > 0
        result

    let CheckUser (userId: string) = 
        let result = database.Table<FacebookProfileDbObject>().Where(w => w.UserId == userId).Count() > 0
        result

    let CheckPostUser (userId: string) =
        let result = database.Table<FacebookPostsDbObject>().Where(userId).Count() > 0
        query {
            for UserId
        }
        let result = data
        result*)

    let DeleteToken =
        database.DeleteAll<TokenDbObject>() |> ignore
        

File: MyFbApp/MyFbAppLibrary/Sqlite/DatabaseManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using MyFbAppLib.DataBaseModels;
using MyFbAppLib.Model;
using SQLite;

namespace MyFbAppLib.Sqlite
{

    public class DatabaseManager
    {
        private SQLiteConnection _dbConnection = new SQLiteConnection(Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.Personal), "MyDB.db3"));

        public DatabaseManager()
        {
        }

        public void CreateTables()
        {
            _dbConnection.CreateTable<FacebookProfileDb>();
            _dbConnection.CreateTable<FacebookPostsDb>();
            _dbConnection.CreateTable<TokenDb>();
        }

        public bool CheckToken()
        {
            return (_dbConnection.Table<TokenDb>().Count() > 0);
        }

        public string getLongLiveToken()
        {
            return (_dbConnection.Table<TokenDb>().First().LongLiveToken);
        }

        public void UpdateLongLiveToken(string token)
        {
            _dbConnection.DeleteAll<TokenDb>();
            var maxPrimaryKey = _dbConnection.Table<TokenDb>().OrderByDescending(c => c.Id).FirstOrDefault();
            TokenDb tokens = new TokenDb
            {
                Id = (maxPrimaryKey == null ? 1 : maxPrimaryKey.Id + 1),
                LongLiveToken =token
            };
            _dbConnection.Insert(tokens);
        }

        public void UpdateProfile(FacebookProfile profile)
        {
            if (CheckUser(profile.Id))
                _dbConnection.Execute("DELETE FROM FacebookProfileDb WHERE UserId = ?", profile.Id);
            var maxPrimaryKey = _dbConnection.Table<FacebookProfileDb>().OrderByDescending(c => c.Id).FirstOrDefault();
            FacebookProfileDb dbprofile = new FacebookProfileDb()
            {
                Id = (maxPrimaryKey == null ? 1 : maxPrimaryKey.Id + 1),
                Name = profile.Name,
                TimeStamp = DateTime.Now,
                UserId = profile.Id
            };

            _dbConnection.Insert(dbprofile);
        }

        public void UpdatePosts(List<PostsData> posts)
        {
            string UserId = posts[0].Id.Substring(0, 16);
            if (CheckPostUser(UserId))
                _dbConnection.Execute("DELETE FROM FacebookPostsDb WHERE UserId = ?", UserId);
            foreach (PostsData post in posts)
            {
                var maxPrimaryKey = _dbConnection.Table<FacebookPostsDb>().OrderByDescending(c => c.Id).FirstOrDefault();
                FacebookPostsDb dbposts = new FacebookPostsDb()
                {
                    Id = (maxPrimaryKey == null ? 1 : maxPrimaryKey.Id + 1),
                    Message = post.Message,
                    TimeStamp = DateTime.Now,
                    CommentsNumber = post.CommentsNumber,
                    PostsId = post.Id,
                    Created_time = post.Created_time,
                    UserId = UserId
                };
                _dbConnection.Insert(dbposts);
            }
        }

        public async Task<FacebookProfileDb> getFacebookProfileDb()
        {
            FacebookProfileDb fbProfile = await Task.Run(() => _dbConnection.Table<FacebookProfileDb>().First());
            return fbProfile;
        }

        public async Task<List<FacebookPostsDb>> getFacebookPostsDb()
        {
            List<FacebookPostsDb> fbPosts = await Task.Run(() => _dbConnection.Table<FacebookPostsDb>().ToList());
            return fbPosts;
        }

        public bool CheckTimeStamp(string UserId)
        {
            return ((DateTime.Compare(_dbConnection.Table<FacebookProfileDb>().Where(w => w.UserId == UserId).First().TimeStamp, DateTime.Now.AddDays(-5)) > 0));
        }

        public bool CheckUser(string UserId)
        {
            return (_dbConnection.Table<FacebookProfileDb>().Where(w => w.UserId == UserId).Count() > 0);
        }

        public bool CheckPostUser(string UserId)
        {
            return (_dbConnection.Table<FacebookPostsDb>().Where(w => w.UserId == UserId).Count() > 0);
        }

        public void DeleteToken()
        {
            _dbConnection.DeleteAll<TokenDb>();
        }

    }
}
